### PR TITLE
User authentication

### PR DESCRIPTION
**Pull request by sparkbrains**

### CODE REVIEW

File: accounts/admin.py [hunk 1]
```diff
@@ -1,11 +1,18 @@
 from django.contrib import admin
-from .models import CustomUser
+from .models import CustomUser, ResetPasswordLink
 
 # Register your models here.
 
 
-class UserAdmin(admin.ModelAdmin):
-    list_display = ("id", "email", "userId")
+class UserProfile(admin.ModelAdmin):
```
Comment (sparkbrains) [inline accounts/admin.py:7]: UserProfile doesn't exists at all. Please change naming conventions properly.

File: accounts/admin.py [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
 
-admin.site.register(CustomUser)
+admin.site.register(CustomUser, UserProfile)
+
+
+class Resetlink(admin.ModelAdmin):
+    list_display = ("id", "user", "created_at", "expired_at")
+
+
+admin.site.register(ResetPasswordLink, Resetlink)
```
Comment (sparkbrains) [inline accounts/admin.py:18]: You haven't ran reformat command. Please run the command @sushil-sbrains

File: accounts/urls.py [hunk 3]
```diff
@@ -1,29 +1,29 @@
...
     path("register/", UserRegistrationView.as_view(), name="register"),
-    # path("update-profile/<int:pk>/", UserProfileUpdate.as_view(), name="update_profile"),
     path(
         "send-password-reset-link/",
         SendPasswordEmailView.as_view(),
         name="password-reset-email.html",
     ),
     path(
-        "password-reset/<str:uuid>",
+        "reset/change-password/<uid>/",
```
Comment (sparkbrains) [inline accounts/urls.py:19]: @sushil-sbrains 
Typecasting is missing from this url. Resolve that. Check the documentation.

https://docs.djangoproject.com/en/4.1/topics/http/urls/#example

File: accounts/views.py [hunk 4]
```diff
@@ -1,41 +1,97 @@
...
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        user_register_obj = serializer.save()
+        recipient_list = str(user_register_obj)
+        context = {
+            "reset_password_obj": user_register_obj.parent_name,
+            "request": request._request,  # WSGI Request object
+        }
+        body_text_content = render_to_string("emails/text/welcome-page.txt")
+        subject = "Welcome to Our Site!"
```
Comment (sparkbrains) [inline accounts/views.py:28]: @sushil-sbrains 


Create one file(i.e constants.py) which includes all the static values. Like subject here.

File: accounts/views.py [hunk 5]
```diff
@@ -1,41 +1,97 @@
...
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        reset_password_obj = serializer.save()
+        context = {
+            "link": settings.FRONT_END_URL,
+            "reset_password_obj": reset_password_obj,
+            "request": request._request,  # WSGI Request object
+        }
+        body_text_content = render_to_string("emails/text/password-reset-email.txt")
+        subject = "Reset Password"
```
Comment (sparkbrains) [inline accounts/views.py:55]: @sushil-sbrains 
Move this to constants.py and then use it.

File: pet/models.py [hunk 6]
```diff
@@ -34,23 +48,37 @@ class PetProfile(models.Model):
...
+        max_length=250, null=True, blank=True, choices=PetGenderChoices.choices
+    )
     height = models.CharField(max_length=250, null=True, blank=True)
     weight = models.CharField(max_length=250, null=True, blank=True)
+    slug = models.SlugField(unique=True)
+
+    def save(self, *args, **kwargs):
+        super(PetProfile, self).save()
+        if not self.slug:
+            N = 4
```
Comment (sparkbrains) [inline pet/models.py:69]: Move this to constants.py

File: pet/serializers.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
 from rest_framework import serializers
 from pet.models import *
+from django.db.models import Q
```
Comment (sparkbrains) [inline pet/serializers.py:3]: I don't see any usage of this operator. Please remove this if you are not using this.

File: templates/emails/html/password-reset-email.html [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+Lorem Ipsum donor Lorem Ipsum donor Lorem Ipsum donor Lorem Ipsum donor Lorem Ipsum donor
+{% endblock %}
+
+{% block button_link %}
+https://google.com
+{% endblock %}
+
+{% block button_text %}
+Reset Password
+{% endblock %}
```
Comment (sparkbrains) [inline templates/emails/html/password-reset-email.html:17]: Please add general html content. Like an actual email is being sent.

File: templates/emails/text/password-reset-email.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+Forgot your password?
+
+Please copy the link below in new tab of your browser:
+
+
+https://google.com
```
Comment (sparkbrains) [inline templates/emails/text/password-reset-email.txt:6]: The whole content is wrong. It should match the content with html file.

File: accounts/admin.py (hunk 1 above)
Comment (sushil-sbrains) [reply accounts/admin.py:7]: Name changed to Customuser from UserProfile in reference to model

File: accounts/admin.py (hunk 2 above)
Comment (sushil-sbrains) [reply accounts/admin.py:18]: resolved

File: accounts/urls.py (hunk 3 above)
Comment (sushil-sbrains) [reply accounts/urls.py:19]: resolved

File: accounts/views.py (hunk 4 above)
Comment (sushil-sbrains) [reply accounts/views.py:28]: resolved

File: accounts/views.py (hunk 5 above)
Comment (sushil-sbrains) [reply accounts/views.py:55]: resolved

File: pet/models.py (hunk 6 above)
Comment (sushil-sbrains) [reply pet/models.py:69]: resolved

File: pet/serializers.py (hunk 7 above)
Comment (sushil-sbrains) [reply pet/serializers.py:3]: resolved

File: accounts/admin.py (hunk 2 above)
Comment (sparkbrains) [reply accounts/admin.py:18]: Customuser change this to CustomUserAdmin

in current scenario CustomUser is a model also and you are naming Admin class as Customuser too. Not appropriate

File: pet/constants.py [hunk 10]
```diff
@@ -0,0 +1 @@
+STRING_RANGE = 4
```
Comment (sparkbrains) [inline pet/constants.py:1]: Change this to appropriate naming conventions.

File: pet/models.py [hunk 11]
```diff
@@ -1,21 +1,36 @@
 from django.db import models
 from accounts.models import CustomUser
 from django.utils.text import slugify
+import string
+import random
+import uuid
+from pet.constants import *
```
Comment (sparkbrains) [inline pet/models.py:7]: Absolute import rather than relative.

File: accounts/constants.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+WELCOME_SUBJECT = "Welcome to Our Site!"
+
+RESET_PASSWORD_SUBJECT = "Reset Password"
```
Comment (sparkbrains) [inline accounts/constants.py:3]: If these variables are being used in sending emails then you can change variable names to 

WELCOME_EMAIL_SUBJECT

Comment (sushil-sbrains) [reply accounts/constants.py:3]: resolved

File: pet/models.py (hunk 11 above)
Comment (sushil-sbrains) [reply pet/models.py:7]: resolved

File: pet/constants.py (hunk 10 above)
Comment (sushil-sbrains) [reply pet/constants.py:1]: resolved

**Pull request merged**